filecoin-proofs/src/caches.rs

```diff
@@ -27,14 +28,74 @@ type Bls12VerifierSRSKey = groth16::aggregate::VerifierSRS<Bls12>;
 type Cache<G> = HashMap<String, Arc<G>>;
 type GrothMemCache = Cache<Bls12GrothParams>;
 type VerifyingKeyMemCache = Cache<Bls12PreparedVerifyingKey>;
-type SRSKeyMemCache = Cache<Bls12ProverSRSKey>;
-type SRSVerifierKeyMemCache = Cache<Bls12VerifierSRSKey>;
+
+const FIP0013_MIN_SNARKS: usize = 64;
+const FIP0013_MAX_SNARKS: usize = 8192;
+
+// Note that proofs testing will use values under and over the FIP13
+// min and max, respectively.
+const PROOFS_TESTS_MIN_SNARKS: usize = FIP0013_MIN_SNARKS >> 5;
+const PROOFS_TESTS_MAX_SNARKS: usize = FIP0013_MAX_SNARKS << 1;
+
+const SRS_IDENTIFIER: &str = "srs-key";
+const SRS_VERIFIER_IDENTIFIER: &str = "srs-verifying-key";
 
 lazy_static! {
     static ref GROTH_PARAM_MEMORY_CACHE: Mutex<GrothMemCache> = Default::default();
     static ref VERIFYING_KEY_MEMORY_CACHE: Mutex<VerifyingKeyMemCache> = Default::default();
-    static ref SRS_KEY_MEMORY_CACHE: Mutex<SRSKeyMemCache> = Default::default();
-    static ref SRS_VERIFIER_KEY_MEMORY_CACHE: Mutex<SRSVerifierKeyMemCache> = Default::default();
+    static ref SRS_KEY_MEMORY_CACHE: SRSCache<Bls12ProverSRSKey> =
+        SRSCache::with_defaults(SRS_IDENTIFIER);
+    static ref SRS_VERIFIER_KEY_MEMORY_CACHE: SRSCache<Bls12VerifierSRSKey> =
+        SRSCache::with_defaults(SRS_VERIFIER_IDENTIFIER);
+}
+
+// We have a separate SRSCache type for srs keys since they are cached
+// differently (as a hashmap per type, keyed by identifier consisting
+// of sector size and pow2 num proofs to aggregate)
+#[derive(Debug, Default)]
+pub struct SRSCache<G> {
+    data: HashMap<String, OnceCell<Arc<G>>>,
+}
+
+impl<G> SRSCache<G> {
+    pub fn with_defaults(identifier: &str) -> Self {
+        let mut data = HashMap::new();
+        let mut num_proofs_to_aggregate = PROOFS_TESTS_MIN_SNARKS;
+
+        loop {
+            for sector_size in &PUBLISHED_SECTOR_SIZES {
+                let key = format!(
+                    "STACKED[{}-{}]-{}",
+                    sector_size, num_proofs_to_aggregate, identifier,
+                );
+                trace!("inserting placeholder srs key with hash key {}", key);
+                data.insert(key.to_string(), OnceCell::new());
+            }
+
+            num_proofs_to_aggregate <<= 1;
+            if num_proofs_to_aggregate > PROOFS_TESTS_MAX_SNARKS {
+                break;
+            }
+        }
+
+        Self { data }
+    }
+
+    /// Returns `None` for non existent entries, `Some(v)` for existing ones, where `v` is either
+    /// the result of running `generator` or already existing one.
+    pub fn get_or_init<F>(&self, key: &str, generator: F) -> Option<&Arc<G>>
+    where
+        F: FnOnce() -> Result<G>,
+    {
+        if let Some(cell) = self.data.get(key) {
+            trace!("generating or waiting on specialize for {}", key);
+            return Some(
+                cell.get_or_init(|| Arc::new(generator().expect("SRS specialize failed"))),
```

Comment:
In which cases can this fail? It seems risky to panic here

Reply:
probably better to change the return value to `Result<Option<&Arc<G>>>>`, and propagate it

Reply:
I'm having trouble propagating the error because it's inside of the generator closure.  It doesn't seem to be happy allowing me to return a Result.

Reply:
this works for me

```rust
    pub fn get_or_init<F>(&self, key: &str, generator: F) -> Result<Option<&Arc<G>>>
    where
        F: FnOnce() -> Result<G>,
    {
        if let Some(cell) = self.data.get(key) {
            trace!("generating or waiting on specialize for {}", key);
            let x = cell.get_or_try_init(|| -> Result<Arc<G>> { Ok(Arc::new(generator()?)) })?;
            return Ok(Some(x));
        }

        Ok(None)
    }
```

Reply:
Ah, I see, thanks.  I had this:

```rust
    pub fn get_or_init<F>(&self, key: &str, generator: F) -> Result<Option<&Arc<G>>>
    where
        F: FnOnce() -> Result<G>,
    {   
        if let Some(cell) = self.data.get(key) {
            trace!("generating or waiting on specialize for {}", key);
            let result = cell.get_or_init(|| -> Result<Arc<G>> {
                let generated = generator()?;
                
                Ok(Arc::new(generated))
            });
            
            return Ok(Some(result));
        }
        
        Ok(None)
    }

```

Which errors with this:

```
error[E0271]: type mismatch resolving `<[closure@filecoin-proofs/src/caches.rs:94:43: 98:14] as FnOnce<()>>::Output == Arc<G>`
  --> filecoin-proofs/src/caches.rs:94:31
   |
94 |             let result = cell.get_or_init(|| -> Result<Arc<G>> {
   |                               ^^^^^^^^^^^ expected enum `std::result::Result`, found struct `Arc`
   |
   = note: expected enum `std::result::Result<Arc<_>, anyhow::Error>`
            found struct `Arc<_>`
```